### PR TITLE
Allow multiple vendors per framework

### PR DESCRIPTION
**Pull request by weierophinney**
First off, I'm very impressed with this project, and have been since I first started seeing you aggregate stats 2 or 3 years ago.
They very closely tracked how we were already tracking our stats for ZF, which gave us high confidence in your comparisons.

As I noted on Twitter, I feel to get an accurate summary of Laminas and ZF, you need to do an aggregate from each of the suborganizations each has/had:

- Laminas:
  - laminas
  - laminas-api-tools
  - mezzio
- ZF:
  - zendframework
  - zfcampus

This patch updates the configuration to add the laminas-api-tools, mezzio, and zfcampus vendors, as well as modify some of the exclusions — I've excluded all skeleton applications, removed the Expressive exclusion (as it was an official subproject of the framework), and added Apigility (the zfcampus repos), Mezzio (formerly Expressive), and Laminas API Tools (formerly Apigility).

Since Laminas API Tools, Mezzio, and Expressive are subprojects of Laminas and Zend Framework, I needed to allow grouping them together for purposes of statistics and reporting.
As such, this patch also provides functionality to allow merging multiple vendors in order to report them under a single framework name.

When run, the landing page continues to report Laminas and Zend, but clicking through to their package pages, you will now see all qualifying packages across all vendor namespaces for those frameworks.
Since package links use the full package name to resolve, all package links continue to resolve correctly.
Moreover, since the Laminas and ZF packages use package prefixes, they are easily recognizable.

To accomplish the merging operations, I moved the calculation of the last year trend percentage to the constructor of `VendorData`.
This allowed me to remove any logic duplication, as the value is derived from other values passed to the constructor.

One potential improvement is to include the full package name, including vendor prefix, on the packages page; I'm willing to make the changes to the templates if this is desired.

### CODE REVIEW

**Comment by weierophinney** [comment]
It looks like the static site builder is not happy with the changes: https://github.com/TomasVotruba/phpfwtrends.org/pull/11/checks#step:5:13

I have ideas about how to resolve that, but will have to wait until tomorrow (2021-03-19) to resolve.

**Comment by weierophinney** [comment]
Okay, all resolved!

The one remaining Rector error is unrelated, and I cannot reproduce it when I run the check locally:

> Rector\BetterPhpDocParser\ValueObject\PhpDocNode\PHPUnit\PHPUnitDataProviderTagValueNode contains 1 abstract method and must therefore be declared abstract or implement the remaining methods (PHPStan\PhpDocParser\Ast\Node::hasAttribute) in /home/runner/work/phpfwtrends.org/phpfwtrends.org/vendor/rector/rector/packages/better-php-doc-parser/src/ValueObject/PhpDocNode/PHPUnit/PHPUnitDataProviderTagValueNode.php on line 11

**Comment by weierophinney** [comment]
All issues resolved!

File: src/ControllerWithDataProvider/FrameworkControllerWithDataProvider.php [hunk 1]
```diff
@@ -32,6 +32,33 @@ public function getControllerMethod(): string
...
         $frameworksVendorToName = $this->parameterProvider->provideArrayParameter(Option::FRAMEWORKS_VENDOR_TO_NAME);
+        $mergeableVendorList    = $this->parameterProvider->provideArrayParameter(Option::FRAMEWORKS_MERGEABLE_VENDORS);
+        $excludableVendors      = array_reduce(
+            $mergeableVendorList,
+            fn (array $excludableVendors, array $currentList): array => $this->reduceExcludableVendors(
+                $excludableVendors,
+                $currentList
+            ),
+            []
+        );
```
Comment (TomasVotruba) [inline src/ControllerWithDataProvider/FrameworkControllerWithDataProvider.php:43]: I cannot understand what this code does. Could you rephrase is so it's self descriptive with less nesting?

File: src/Merger.php [hunk 2]
```diff
@@ -0,0 +1,70 @@
+<?php
+
+declare(strict_types=1);
+
+namespace TomasVotruba\PhpFwTrends;
+
+use TomasVotruba\PhpFwTrends\ValueObject\VendorData;
+use Webmozart\Assert\Assert;
+
+final class Merger
```
Comment (TomasVotruba) [inline src/Merger.php:10]: This class needs tests.

"Merger" could be related to anything between 2 items. More precise name would help to understand too.

File: src/MergeableVendors.php [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+declare(strict_types=1);
+
+namespace TomasVotruba\PhpFwTrends;
+
+use Symplify\PackageBuilder\Parameter\ParameterProvider;
+use TomasVotruba\PhpFwTrends\Exception\ShouldNotHappenException;
+use TomasVotruba\PhpFwTrends\ValueObject\Option;
+use Webmozart\Assert\Assert;
+
+final class MergeableVendors
```
Comment (TomasVotruba) [inline src/MergeableVendors.php:12]: Please add a test for this one

**Comment by TomasVotruba** [comment]
Thanks for you work :+1: 

I've added few comments

File: src/ControllerWithDataProvider/FrameworkControllerWithDataProvider.php (hunk 1 above)
Comment (weierophinney) [reply src/ControllerWithDataProvider/FrameworkControllerWithDataProvider.php:43]: Essentially, I'm using the return value of a function call to accomplish a reduction process. The reduction process itself is used to identify vendors that should be excluded from those reported on the home page, due to the fact that their data will be merged under another entry.

I see two ways to clarify what is happening here:

- Inline the function itself (`FrameworkControllerWithDataProvider::reduceExcludableVendors`) here.
- Move the reduction operation into a new function entirely, something along the line of `removeExcludableVendorsFromFrameworkVendorName(array $frameworksVendorToName): array`.

Which would you prefer? Or do you have another suggestion?

Comment (TomasVotruba) [reply src/ControllerWithDataProvider/FrameworkControllerWithDataProvider.php:43]: Thanks. This comment should be embodied in the code itself, so it's not needed to explain it.

> Move the reduction operation into a new function entirely.

This sound good :+1: It can be covered with test as well then.

**Comment by weierophinney** [comment]
@TomasVotruba I've performed refactors per your request, and rebased to pick up more recent changes.

**Review by TomasVotruba — APPROVED** [review]

**Comment by TomasVotruba** [comment]
Thank you!

Looks good, let's ship it :)
